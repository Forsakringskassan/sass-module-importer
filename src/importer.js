import path from "node:path";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";
import { createRequire } from "node:module";

import resolvePackagePath from "resolve-package-path";
import { exports, legacy } from "resolve.exports";

import { getPackageNameFromPath } from "./parsePackageName";

const { findUpPackagePath } = resolvePackagePath;
const require = createRequire(import.meta.url);
const WEBPACK_NODE_MODULE_PREFIX = "~";

let selfPackageJson = null;
let selfPackageJsonPath = null;
export const moduleImporter = {
    findFileUrl(url) {
        setSelfPackage();

        let findUrl = url;
        if (url.startsWith(WEBPACK_NODE_MODULE_PREFIX)) {
            findUrl = url.substring(1);
        }

        const packageName = getPackageNameFromPath(findUrl);
        const filePath = findUrl.replace(packageName, "");

        /* Validate if packageName is valid */
        if (!packageName) {
            return null;
        }

        let packageJson = selfPackageJson;
        let packagePath = selfPackageJsonPath;

        if (selfPackageJson.name !== packageName) {
            packagePath = resolvePackagePath(packageName, process.cwd());

            /* Validate if existing package */
            if (!packagePath) {
                return null;
            }

            packageJson = JSON.parse(
                readFileSync(packagePath, { encoding: "utf-8" }),
            );
        }

        const moduleDirectory = path.dirname(packagePath);

        /* Check exports */
        try {
            const match = exports(packageJson, filePath.substring(1), {
                conditions: ["sass"],
            });
            if (match && match.length === 1) {
                return new URL(
                    pathToFileURL(path.join(moduleDirectory, match[0])),
                );
            }
        } catch {
            /* empty */
        }

        /* Check main fields (only applies if only package path is given) */
        if (!filePath) {
            const match = legacy(packageJson, { fields: ["sass", "main"] });
            if (match) {
                return new URL(
                    pathToFileURL(path.join(moduleDirectory, match)),
                );
            }
        }

        /* Direct link */
        const directory = path.dirname(filePath);
        const fileName = path.basename(filePath);

        const search = [
            `${fileName}.css`,
            `${fileName}.scss`,
            `_${fileName}.scss`,
            `${fileName}`,
            `${fileName}/_index.scss`,
        ];

        for (const variant of search) {
            try {
                const moduleName = path.posix.join(
                    moduleDirectory,
                    directory,
                    variant,
                );
                const resolved = require.resolve(moduleName);
                return new URL(pathToFileURL(resolved));
            } catch (err) {
                if (err.code !== "MODULE_NOT_FOUND") {
                    throw err;
                }
            }
        }
        return null;
    },
};

function setSelfPackage() {
    if (!selfPackageJson) {
        selfPackageJsonPath = findUpPackagePath(process.cwd());
        selfPackageJson = JSON.parse(
            readFileSync(selfPackageJsonPath, { encoding: "utf-8" }),
        );
    }
}
