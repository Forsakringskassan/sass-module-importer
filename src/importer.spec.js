import fs from "node:fs";
import { compileString } from "sass";
import { moduleImporter } from "./importer";

it("should be able to transform scss", () => {
    const result = compileString(
        fs.readFileSync("./fixtures/main.scss", "utf-8"),
        {
            style: "expanded",
            importers: [moduleImporter],
        },
    );
    expect(result.css).toMatchSnapshot();
});
