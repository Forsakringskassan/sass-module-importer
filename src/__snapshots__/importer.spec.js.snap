// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to transform scss using package with exports 1`] = `
"span {
  color: pink;
}

body {
  color: orange;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package with main field 1`] = `
".main {
  color: blue;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package with sass fields 1`] = `
".main {
  color: green;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss using package without exports and main fields 1`] = `
"html,
body {
  background-color: #12c344;
}

.bar {
  color: greenyellow;
}

.reset {
  color: inherit;
}

.foo {
  color: green;
}"
`;

exports[`should be able to transform scss when package contains reference to itself 1`] = `
".self {
  color: inherit;
}

.foo {
  color: green;
}"
`;
